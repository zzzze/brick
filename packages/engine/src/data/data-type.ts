export interface TypeValidator {
  (value: unknown): boolean
}

export interface DataType<T = unknown> {
  type: string
  default: T
  isValid: TypeValidator
}

export type DataConfig = Record<string, DataType>

export const StringType: DataType<string> = {
  type: 'string',
  default: '',
  isValid: (value) => typeof value === 'string',
}

export const NumberType: DataType<number> = {
  type: 'number',
  default: 0,
  isValid: (value) => typeof value === 'number' && value == value,
}

export const BooleanType: DataType<boolean> = {
  type: 'boolean',
  default: false,
  isValid: (value) => typeof value === 'boolean',
}
