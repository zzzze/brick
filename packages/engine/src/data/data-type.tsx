import React, { ReactElement } from 'react'
import { ObjectStringInput } from '@brick/components'
import isPlainObject from 'lodash/isPlainObject'

export interface TypeValidator {
  (value: unknown): boolean
}

export interface DataType<T = unknown> {
  type: string
  default: T
  label?: string
  placeholder?: string
  testID?: string
  formItem: () => ReactElement
  isValid: TypeValidator
}

export type DataConfig = Record<string, DataType>

export const StringType: DataType<string> = {
  type: 'string',
  default: '',
  formItem: (placeholder?: string) => <input type="text" placeholder={placeholder} />,
  isValid: (value) => typeof value === 'string',
}

export const NumberType: DataType<number> = {
  type: 'number',
  default: 0,
  formItem: (placeholder?: string) => <input type="text" placeholder={placeholder} />,
  isValid: (value) => typeof value === 'number' && value == value,
}

export const BooleanType: DataType<boolean> = {
  type: 'boolean',
  default: false,
  formItem: (placeholder?: string) => <input type="text" placeholder={placeholder} />,
  isValid: (value) => typeof value === 'boolean',
}

export const ObjectType: DataType<Record<string, unknown>> = {
  type: 'object',
  default: {},
  formItem: () => <ObjectStringInput />,
  isValid: (value) => isPlainObject(value),
}

export const CodeType: DataType<string> = {
  type: 'code',
  default: '',
  formItem: () => <textarea />,
  isValid: (value) => typeof value === 'string',
}
