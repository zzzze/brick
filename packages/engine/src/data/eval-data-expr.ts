export default function evalExpr(
  expression: string,
  $this: Record<string, unknown>,
  pSupply: Record<string, unknown>
): unknown {
  expression = expression.replace(/^\{\{|\}\}$/g, '')
  expression = expression.replace(/(\$?\w?[a-zA-Z]+\w?\.\w)/g, (match) =>
    !match.startsWith('$this.') ? `pSupply.${match}` : match
  )
  return new Function('$this', 'pSupply', 'return ' + expression)($this, pSupply) // eslint-disable-line
}
