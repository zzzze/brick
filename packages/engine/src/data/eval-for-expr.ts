export default function evalForExpr(
  expression: string,
  $this: Record<string, unknown>,
  item: unknown,
  index: number
): unknown {
  expression = expression.replace(/^\{\{|\}\}$/g, '')
  return new Function('$this', 'item', 'index', 'return ' + expression)($this, item, index) // eslint-disable-line
}
