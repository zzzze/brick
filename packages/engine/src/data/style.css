.formitem-input {
  border: solid 1px #ccc;
  border-radius: 5px;
  padding: 5px;
  flex: 1;
}

.formitem-textarea {
  border: solid 1px #ccc;
  border-radius: 5px;
  padding: 5px;
  flex: 1;
  min-height: 100px;
  resize: none;
}
