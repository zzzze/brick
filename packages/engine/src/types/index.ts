export * from './brick'
export * from './config'
export * from './engine'
export * from './brick-instance'
