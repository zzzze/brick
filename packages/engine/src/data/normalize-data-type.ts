import { ReactElement } from 'react'
import { DataType } from './data-type'

export interface DataTypeDefinition<T = unknown> extends Omit<DataType<T>, 'isValid' | 'formItem'> {
  formItem?: () => ReactElement
}

export function checkTypeIsString(dataType: string | DataTypeDefinition): dataType is string {
  return typeof dataType === 'string'
}

export default function normalizeDataType(
  registeredTypes: Record<string, DataType<unknown>>,
  dataTypes: Record<string, string | DataTypeDefinition>
): Record<string, DataType> {
  return Object.keys(dataTypes).reduce<Record<string, DataType>>((result, key) => {
    const typeDefinition = dataTypes[key]
    if (checkTypeIsString(typeDefinition)) {
      if (!registeredTypes[typeDefinition]) {
        throw Error(`type "${typeDefinition}" not found`)
      }
      result[key] = {
        ...registeredTypes[typeDefinition],
      }
    } else {
      if (!registeredTypes[typeDefinition.type]) {
        throw Error(`type "${typeDefinition.type}" not found`)
      }
      result[key] = {
        ...registeredTypes[typeDefinition.type],
        ...typeDefinition,
      }
    }
    return result
  }, {})
}
