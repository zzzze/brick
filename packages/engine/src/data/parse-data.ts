import { DataObject } from '../types'
import { DataConfig } from './data-type'
import { ContextType } from '../context'
import evalExpr from './eval-data-expr'
import { captureAttrDependencies, isExpression, isForParamsExpression } from '../utils/expression'

/**
 * Parse data of brick instance
 * @param {DataConfig} dataConfig data config from brick definition
 * @param {DataObject} data data of brick instance
 * @param {DataObject} supply data from parent brick
 * @returns {DataObject}
 * @example
 *    dataConfig:
 *    {
 *      name: {
 *         type: 'string',
 *         default: 'foo',
 *      },
 *      age: {
 *         type: 'number',
 *         default: 1,
 *      },
 *    }
 *
 *    pSupply:
 *    {
 *        $container: {
 *            age: 10,
 *        },
 *    }
 *
 *    data:
 *    {
 *        name: 'bar',
 *        nickname: '{{$this.name}}',
 *        age: '{{$container.age}}',
 *    }
 *
 *    =====>
 *
 *    {
 *        name: 'bar',
 *        nickname: 'bar',
 *        age: 10,
 *    }
 */
export default function parseData(
  engineCtx: ContextType,
  dataConfig: DataConfig,
  data: DataObject,
  pSupply: DataObject
): DataObject {
  const keys = Array.from(new Set([...Object.keys(dataConfig), ...Object.keys(data)]))
  const newData = keys.reduce<DataObject>((result, key) => {
    if (typeof result[key] !== 'undefined') {
      return result
    }
    const getValueOfkey = (result: DataObject, key: string, traversedKeys: string[]) => {
      if (traversedKeys.includes(key)) {
        throw Error('circular dependence')
      }
      traversedKeys = traversedKeys.concat(key)
      let value = data[key] ?? dataConfig[key].default
      if (
        isExpression(value, engineCtx.options.delimiters) &&
        !isForParamsExpression(value, engineCtx.options.delimiters, [
          engineCtx.options.identifiers.forItem,
          engineCtx.options.identifiers.forIndex,
        ])
      ) {
        const matches = captureAttrDependencies(value, `${engineCtx.options.identifiers.idPrefix}this`)
        const dependentData: DataObject = {}
        if (matches) {
          matches.forEach((attr) => {
            const valueOfAttr = getValueOfkey(result, attr, traversedKeys)
            dependentData[attr] = valueOfAttr
          })
        }
        value = evalExpr(value, { ...dependentData }, pSupply)
      }
      // TODO: validate value
      result[key] = value
      return value
    }
    const value = getValueOfkey(result, key, [])
    result[key] = value
    return result
  }, {})
  return newData
}
