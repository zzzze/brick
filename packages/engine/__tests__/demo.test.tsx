import React from 'react'
import demoTest from '@brick/tools/tests/demoTest'
import { Engine } from '@/index'
import { BrickContainer } from '@brick/components'
import { Brick, ChildrenType, BrickInstance } from '@/types'
import { ConfigurationFormItem as FormItem } from '@brick/components'
import { StringType, NumberType, BooleanType } from '../src/data/data-type'

const View: Brick = {
  name: 'View',
  dataTypes: {
    name: 'string',
  },
  childrenType: ChildrenType.MULTIPLE,
  renderConfigForm() {
    return <></>
  },
  render(args: BrickInstance) {
    return <BrickContainer>{args.children}</BrickContainer>
  },
  version: '0.0.1',
}

const Text: Brick = {
  name: 'Text',
  dataTypes: {
    content: 'string',
  },
  childrenType: ChildrenType.NONE,
  renderConfigForm() {
    return (
      <FormItem label="edit Text: " name="content">
        <input data-testid="content-input" />
      </FormItem>
    )
  },
  render(args: BrickInstance) {
    return <BrickContainer tag="span">{args.data.content as string}</BrickContainer>
  },
  version: '0.0.1',
}

const ViewWithCustomRender: Brick = {
  name: 'ViewWithCustomRender',
  dataTypes: {},
  childrenType: ChildrenType.MULTIPLE,
  canCustomizeRender: true,
  renderConfigForm() {
    return <></>
  },
  render(args: BrickInstance) {
    void args
    return <></>
  },
  version: '0.0.1',
}

Engine.registerBrick(View)
Engine.registerBrick(Text)
Engine.registerBrick(ViewWithCustomRender)

Engine.registerDataType(StringType)
Engine.registerDataType(NumberType)
Engine.registerDataType(BooleanType)

demoTest('engine')
