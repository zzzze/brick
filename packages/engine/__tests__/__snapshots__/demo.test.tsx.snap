// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ./packages/engine/demo/basic.md correctly 1`] = `
<span
  class="brick brick__with-config-form"
>
  <div
    class="brick__config-form"
  >
    <button
      class="brick__config-form-btn brick__config-form-btn-edit"
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  hello world
</span>
`;

exports[`renders ./packages/engine/demo/children.md correctly 1`] = `
<div
  class="brick brick__with-config-form"
>
  <div
    class="brick__config-form"
  >
    <button
      class="brick__config-form-btn brick__config-form-btn-edit"
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <span
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    hello
  </span>
  <span
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    world
  </span>
</div>
`;

exports[`renders ./packages/engine/demo/config-tree-array.md correctly 1`] = `
Array [
  <span
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    hello
  </span>,
  <span
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    world
  </span>,
  <div
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span
      class="brick brick__with-config-form"
    >
      <div
        class="brick__config-form"
      >
        <button
          class="brick__config-form-btn brick__config-form-btn-edit"
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      hello
    </span>
    <span
      class="brick brick__with-config-form"
    >
      <div
        class="brick__config-form"
      >
        <button
          class="brick__config-form-btn brick__config-form-btn-edit"
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      world
    </span>
  </div>,
]
`;

exports[`renders ./packages/engine/demo/custom-render.md correctly 1`] = `
<div
  class="brick brick__with-config-form"
>
  <div
    class="brick__config-form"
  >
    <button
      class="brick__config-form-btn brick__config-form-btn-edit"
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <span>
    foo
  </span>
  <span
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    bar
  </span>
</div>
`;

exports[`renders ./packages/engine/demo/supply.md correctly 1`] = `
<div
  class="brick brick__with-config-form"
>
  <div
    class="brick__config-form"
  >
    <button
      class="brick__config-form-btn brick__config-form-btn-edit"
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <div
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span
      class="brick brick__with-config-form"
    >
      <div
        class="brick__config-form"
      >
        <button
          class="brick__config-form-btn brick__config-form-btn-edit"
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      foo
    </span>
  </div>
</div>
`;

exports[`renders ./packages/engine/demo/supply-from-data.md correctly 1`] = `
<div
  class="brick brick__with-config-form"
>
  <div
    class="brick__config-form"
  >
    <button
      class="brick__config-form-btn brick__config-form-btn-edit"
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <div
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span
      class="brick brick__with-config-form"
    >
      <div
        class="brick__config-form"
      >
        <button
          class="brick__config-form-btn brick__config-form-btn-edit"
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      bar
    </span>
  </div>
</div>
`;

exports[`renders ./packages/engine/demo/supply-with-id.md correctly 1`] = `
<div
  class="brick brick__with-config-form"
>
  <div
    class="brick__config-form"
  >
    <button
      class="brick__config-form-btn brick__config-form-btn-edit"
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <div
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span
      class="brick brick__with-config-form"
    >
      <div
        class="brick__config-form"
      >
        <button
          class="brick__config-form-btn brick__config-form-btn-edit"
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      foo
    </span>
  </div>
</div>
`;

exports[`renders ./packages/engine/demo/supply-without-id.md correctly 1`] = `
<div
  class="brick brick__with-config-form"
>
  <div
    class="brick__config-form"
  >
    <button
      class="brick__config-form-btn brick__config-form-btn-edit"
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <div
    class="brick brick__with-config-form"
  >
    <div
      class="brick__config-form"
    >
      <button
        class="brick__config-form-btn brick__config-form-btn-edit"
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span
      class="brick brick__with-config-form"
    >
      <div
        class="brick__config-form"
      >
        <button
          class="brick__config-form-btn brick__config-form-btn-edit"
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      baz
    </span>
  </div>
</div>
`;
